src/charm.py

```diff
@@ -37,24 +45,97 @@ def __init__(self, *args):
+
+        # validate configuration options
+        missing_config = self._check_config()
+        if missing_config:
+            logger.error('Incomplete Configuration : {}. '
+                         'Application will be blocked.'.format(missing_config))
+            self.unit.status = \
+                BlockedStatus('Missing configuration: {}'.format(missing_config))
+            return
+
+        # check if configuration file has changed and if so push the
+        # new config file to the workload container
+        prometheus_config = self._prometheus_config()
+        config_hash = str(hashlib.md5(str(prometheus_config).encode('utf-8')))
+        if not self._stored.prometheus_config_hash == config_hash:
+            self._stored.prometheus_config_hash = config_hash
+            container.push(PROMETHEUS_CONFIG, prometheus_config)
+            logger.info("Pushed new configuration")
+
+        # setup the workload (Prometheus) container and its services
+        layer = self._prometheus_layer()
+        plan = container.get_plan()
+        if plan.services != layer["services"]:
+            container.add_layer("prometheus", layer, combine=True)
+
+            if container.get_service("prometheus").is_running():
+                container.stop("prometheus")
+
+            container.start("prometheus")
+            logger.info("Restarted prometheus container")
+
+        self.app.status = ActiveStatus()
+        self.unit.status = ActiveStatus()
+
+    def _on_update_status(self, event):
+        """Check status of Prometheus server.
+
+        Status of the Prometheus services is checked by querying
+        Prometheus for its version information. If Prometheus responds
+        with valid information, its active status is recorded.
+        """
+        provided = {'prometheus': self.version}
+        if provided['prometheus']:
+            logger.debug("Prometheus provider is available")
+            logger.debug("Providing : {}".format(provided))
+            if not self._stored.provider_ready:
+                self._stored.provider_ready = True
 
     def _on_stop(self, _):
-        """Mark unit is inactive
+        """Mark unit is inactive.
+
+        All units of the charm are set to maintenance status before
+        termination.
         """
         self.unit.status = MaintenanceStatus('Pod is terminating.')
 
     def _on_grafana_changed(self, event):
-        """Provide Grafana with data source information
+        """Provide Grafana with data source information.
+
+        Grafana needs to know the port and name of an application in order
+        to form a relation with it. Hence this information is provided here.
         """
         event.relation.data[self.unit]['port'] = str(self.model.config['port'])
         event.relation.data[self.unit]['source-type'] = 'prometheus'
+        event.relation.data[self.unit]['private-address'] = str(
+            self.model.get_binding(event.relation).network.bind_address
+        )
```

Comment:
does the use of a private address imply that all units are always deployed on the same local network?

Reply:
Good catch. I need to think about this one.

Reply:
The underlying assumption from the inherited code may be that the LMA model running inside k8s will have an overlay network which routes between nodes.

It's probably a field which we should rename, and dynamically decide whether we want to use `.network.bind_address` or an ingress address if there's an ingress configured, but not all charms *should* necessarily have an ingress exposed unless it means something very different for charms than it means for k8s/"floating IPs" in openstack/etc

Reply:
I have created a github [issue ticket](https://github.com/canonical/prometheus-operator/issues/25) for gathering requirements and finalizing design on this issue. I am resolving this conversation and will take this issue up on a future commit, as discussed. Thank you for highlighting the issue.